File: algorithms/sorting/javascript/insertion.js
/*
Insertion sort is in-place sorting algorithm.
Time complexity : O(n^2)
Space complexity : O(1)
*/

/*
insertionSort is a function that takes an array of numbers.
Sort them through insertion sort technique. 
*/

//Author : Parixit Singh

//Sorting the array numbers
var insertionSort = function(numbersArr) {
    //Creating a copy of the array passed to the function
    var numbers = ([]).concat(numbersArr);
    for (var i = 0; i < numbers.length; i++) {
        //Keeping the value at the i position, and i as the position
        var value = numbers[i],
            position = i;
        //Checking the position and comparing the value 
        //Here shifting of the numbers taking place
        //Values that are greater are shifted one index down
        while ((position > 0) && (numbers[position - 1] > value)) {
            numbers[position] = numbers[position - 1];
            position--;
        }
        //Inserting the value at the required position
        numbers[position] = value;
    }

    //Returning the numbers arrays
    return numbers;
};

File: algorithms/sorting/go/Insertion.go
package main

import (
	"fmt"
	"os"
)

func main() {
	var numbers [10]int
	var num int
	for i := 0; i < 10; i++ {
		_, err := fmt.Scanf("%d", &num)
		if err == nil {
			numbers[i] = num
		} else {
			fmt.Printf("Numbers are not space seperated %v \n", err)
			os.Exit(1)
		}
	}

	for i := 0; i < 10; i++ {
		value := numbers[i]
		position := i
		for position > 0 && numbers[position-1] > value {
			numbers[position] = numbers[position-1]
			position--
		}
		numbers[position] = value
	}

	for i := 0; i < 10; i++ {
		fmt.Printf("%d ", numbers[i])
	}
}
